dartsim/src/GzOdeCollisionDetector.cc

```diff
@@ -0,0 +1,112 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#include <memory>
+#include <unordered_map>
+
+#include <dart/collision/CollisionObject.hpp>
+
+#include "GzOdeCollisionDetector.hh"
+
+using namespace dart;
+using namespace collision;
+
+/////////////////////////////////////////////////
+GzOdeCollisionDetector::GzOdeCollisionDetector()
+  : OdeCollisionDetector()
+{
+}
+
+/////////////////////////////////////////////////
+GzOdeCollisionDetector::Registrar<GzOdeCollisionDetector>
+    GzOdeCollisionDetector::mRegistrar{
+        GzOdeCollisionDetector::getStaticType(),
+        []() -> std::shared_ptr<GzOdeCollisionDetector> {
+          return GzOdeCollisionDetector::create();
+        }};
+
+/////////////////////////////////////////////////
+std::shared_ptr<GzOdeCollisionDetector> GzOdeCollisionDetector::create()
+{
+  return std::shared_ptr<GzOdeCollisionDetector>(new GzOdeCollisionDetector());
+}
+
+/////////////////////////////////////////////////
+bool GzOdeCollisionDetector::collide(
+    CollisionGroup *_group,
+    const CollisionOption &_option,
+    CollisionResult *_result)
+{
+  bool ret = OdeCollisionDetector::collide(_group, _option, _result);
+  this->LimitCollisionPairMaxContacts(_result);
+  return ret;
+}
+
+/////////////////////////////////////////////////
+bool GzOdeCollisionDetector::collide(
+    CollisionGroup *_group1,
+    CollisionGroup *_group2,
+    const CollisionOption &_option,
+    CollisionResult *_result)
+{
+  bool ret = OdeCollisionDetector::collide(_group1, _group2, _option, _result);
+  this->LimitCollisionPairMaxContacts(_result);
+  return ret;
+}
+
+/////////////////////////////////////////////////
+void GzOdeCollisionDetector::SetCollisionPairMaxContacts(
+    std::size_t _maxContacts)
+{
+  this->maxCollisionPairContacts = _maxContacts;
+}
+
+/////////////////////////////////////////////////
+std::size_t GzOdeCollisionDetector::GetCollisionPairMaxContacts() const
+{
+  return this->maxCollisionPairContacts;
+}
+
+/////////////////////////////////////////////////
+void GzOdeCollisionDetector::LimitCollisionPairMaxContacts(
+    CollisionResult *_result)
+{
+  if (this->maxCollisionPairContacts ==
+    std::numeric_limits<std::size_t>::max())
+    return;
+
+  auto allContacts = _result->getContacts();
+  _result->clear();
+
+  std::unordered_map<dart::collision::CollisionObject *,
+      std::unordered_map<dart::collision::CollisionObject *, std::size_t>>
+      contactMap;
+
+  for (auto &contact : allContacts)
+  {
+    auto &count =
+      contactMap[contact.collisionObject1][contact.collisionObject2];
+    count++;
+    auto &otherCount =
+      contactMap[contact.collisionObject2][contact.collisionObject1];
+    std::size_t total =  count + otherCount;
+    if (total <= this->maxCollisionPairContacts)
```

Comment:
my only other thought is about how to choose between multiple contact points if we need to drop some, but I see that is what #584 is for, so I'll go look at that now

Reply:
responded to comments in https://github.com/gazebosim/gz-physics/pull/584